hw/formal/tools/jaspergold/parse-fpv-report.py

```diff
@@ -0,0 +1,180 @@
+#!/usr/bin/env python3
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+"""
+Parses JasperGold FPV logfile and coverage and dump filtered messages in hjson format.
+"""
+
+import argparse
+import re
+from pathlib import Path
+from collections import OrderedDict
+
+import hjson
+
+
+def extract_messages(full_file, patterns, results):
+    """
```

Comment:
Your docstrings are slightly unusually formatted. There's a [PEP](https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings) describing the standard way to do things. For this one, I think the text could be improved a bit too. Something like this?

```python
'''Extract messages matching patterns from full_file as a dictionary.

The patterns argument is a list of pairs, (severity, pattern). Each pattern is a
regex and all matches in full_file are stored in a dictionary under the key severity.
'''
```

I also don't think that passing results in is a very clean way to do things. Better is probably to use [`dict.update`](https://docs.python.org/3/library/stdtypes.html#dict.update) like this:

``` python
messages.update(extract_messages(full_file, err_warn_patterns))
```

and define results as an empty `OrderedDict` in the function.

I also wonder about the `full_file` argument name: could this be something like "buffer"? Otherwise, it looks like the object might be something to do with files, rather than strings.

Reply:
I'll look properly at the PR in a minute, but replying to this quickly first.

For (1), I'm not sure. I don't see any strong arguments either way or the other: maybe they're actually just trying to emphasize the "triple-ness" rather than "double quote-ness"? I've definitely seen both in Python codebases, and people seem to lean more towards single quotes (maybe because they are less visually noisy with most fonts?).

For (2), I don't think it really matters. I think the example you pointed at is my code, and that's formatted like that because it's what the Python mode in Emacs does on fill-paragraph (M-q). If there's a more compelling reason for one option or the other, I'm afraid I don't know it!

For (3), it doesn't matter, I don't think. Nice one for finding an inconsistent example in only ~20 lines :-) (which I think might be my code again...) Note that these examples have one string per line (using Python's C-style implicit string concatenation), so they don't have any special rules: they are just short strings laid out between parentheses.

Reply:
I see, thanks for your replies!